include/wlr/xwayland.h

```diff
@@ -66,14 +66,23 @@ struct wlr_xwayland_surface_size_hints {
 
 struct wlr_xwayland_surface {
 	xcb_window_t window_id;
+	struct wlr_xwm *xwm;
 	uint32_t surface_id;
-	struct wl_list link;
+
+	struct wl_list displayable_link;
+	// XXX: I think this is just a list of all the surfaces
+	struct wl_list new_link;
+	struct wl_list unpaired_link;
```

Comment:
The way you handle these three lists seems a bit odd, you never remove anything from the new_link or unpaired_link?
The way the original link worked there was only one link for multiple lists on purpose as it is a state machine. I've already forgotten my old code but basically we're waiting for two events to display the window, something from the X11 side and something from the wayland side (Xwayland is a wayland client to us) ; if one comes first the surface is put in new_surfaces, if the other comes first it's put in unpaired_surfaces, and when you have both whichever came first it would go to displayable_surfaces. The surface cannot be in multiple lists at once.

I don't really care if you want the increased complexity of handling three links + associated bools but at the very least never cleaning the other two lists is going to make things slower than it needs to be when there are many windows around...

Reply:
Oh thanks for clearing that up. I just assumed whoever wrote the code didn't know how `wl_list` worked. I will definitely make this much simpler in the course of my rewrite.